### PR TITLE
Getting root hash for predicates when building

### PR DESCRIPTION
**Pull request by seem-less**
Closes [#1985](https://github.com/FuelLabs/sway/issues/1985).

I am not entirely certain what should be prepended to the root. I added "0x" to it like like regular hashes, but not sure if this is correct.

### CODE REVIEW

File: forc/src/ops/forc_build.rs [hunk 1]
```diff
@@ -170,7 +171,16 @@ pub fn build(command: BuildCommand) -> Result<pkg::Compiled> {
...
+        info!("  Hash of the Script bytecode is: {}", bytecode_hash);
+    }
+
+    if let TreeType::Predicate = compiled.tree_type {
+        // get the root hash of the bytecode for predicates and store the result in a file in the output directory
+        let root = format!("0x{}", Contract::root_from_code(&compiled.bytecode));
+        let root_file_name = format!("{}{}", &manifest.project.name, SWAY_BIN_ROOT_SUFFIX);
+        let root_path = output_dir.join(root_file_name);
+        fs::write(root_path, &root)?;
+        info!("  Root hash of the Predicate bytecode: {}", root);
```
Comment (mohammadfawaz) [inline forc/src/ops/forc_build.rs:183]: How about:
```suggestion
        info!("  Predicate root: {}", root);
```

File: forc/src/utils/mod.rs [hunk 2]
```diff
@@ -11,3 +11,7 @@ pub const SWAY_GIT_TAG: &str = concat!("v", clap::crate_version!());
 /// The suffix that helps identify the file which contains the hash of the binary file created when
 /// scripts are built.
 pub const SWAY_BIN_HASH_SUFFIX: &str = "-bin-hash";
+
+/// The suffix that helps identify the file which contains the root hash of the binary file created
+/// when predicates are built.
+pub const SWAY_BIN_ROOT_SUFFIX: &str = "-bin-root-hash";
```
Comment (mohammadfawaz) [inline forc/src/utils/mod.rs:17]:
```suggestion
pub const SWAY_BIN_ROOT_SUFFIX: &str = "-bin-root";
```

File: forc/src/ops/forc_build.rs [hunk 3]
```diff
@@ -170,7 +171,16 @@ pub fn build(command: BuildCommand) -> Result<pkg::Compiled> {
         let hash_file_name = format!("{}{}", &manifest.project.name, SWAY_BIN_HASH_SUFFIX);
         let hash_path = output_dir.join(hash_file_name);
         fs::write(hash_path, &bytecode_hash)?;
-        info!("  Bytecode hash is: {}", bytecode_hash);
+        info!("  Hash of the Script bytecode is: {}", bytecode_hash);
+    }
+
+    if let TreeType::Predicate = compiled.tree_type {
+        // get the root hash of the bytecode for predicates and store the result in a file in the output directory
+        let root = format!("0x{}", Contract::root_from_code(&compiled.bytecode));
```
Comment (mohammadfawaz) [inline forc/src/ops/forc_build.rs:179]: `0x` here is correct. The root is indeed a hash.

File: forc/src/ops/forc_build.rs [hunk 4]
```diff
@@ -170,7 +171,16 @@ pub fn build(command: BuildCommand) -> Result<pkg::Compiled> {
         let hash_file_name = format!("{}{}", &manifest.project.name, SWAY_BIN_HASH_SUFFIX);
         let hash_path = output_dir.join(hash_file_name);
         fs::write(hash_path, &bytecode_hash)?;
-        info!("  Bytecode hash is: {}", bytecode_hash);
+        info!("  Hash of the Script bytecode is: {}", bytecode_hash);
```
Comment (mohammadfawaz) [inline forc/src/ops/forc_build.rs:174]: Maybe just "Script bytecode hash: {}"?

This is the current format for contracts when running `forc deploy`:
```console
  Bytecode size is 112 bytes.
Contract id: 0x01f0c9a726086b74216f2c5ddaa13378c0636c8c926552bac8234b4a07c8344d
Logs:
TransactionId(HexFormatted(68a19c3ec223d8d276ad112892ec65780e9596b708921be98de78a8fd28486db))
```

**Review by mohammadfawaz — COMMENTED** [review]
Thank you for following up on this! This is great!

File: forc/src/ops/forc_build.rs [hunk 5]
```diff
@@ -170,7 +171,16 @@ pub fn build(command: BuildCommand) -> Result<pkg::Compiled> {
...
+    }
+
+    if let TreeType::Predicate = compiled.tree_type {
```
Comment (mohammadfawaz) [inline forc/src/ops/forc_build.rs:177]: These two consecutive `if let` can probably be replaced with ` match` statement where we do nothing for `_`.

File: forc/src/cli/commands/build.rs [hunk 6]
```diff
@@ -4,12 +4,17 @@ use clap::Parser;
...
+/// The output produced will depend on the project's program type.
 ///
-/// Building contracts and libraries will also produce the public ABI in JSON format
+/// - `script`, `predicate` and `contract` projects will produce their bytecode in binary format `<project-name>.bin`.
+///
+/// - `script` projects will also produce a file containing the hash of the bytecode binary
+/// `<project-name>-bin-hash` (using `fuel_cypto::Hasher`).
+///
+/// - `predicate` projects will also produce a file containing the **root** hash of the bytecode binary
+/// `<project-name>-bin-root-hash` (using `fuel_tx::Contract::root_from_code`).
```
Comment (mitchmindtree) [inline forc/src/cli/commands/build.rs:15]:
```suggestion
/// `<project-name>-bin-root` (using `fuel_tx::Contract::root_from_code`).
```

Comment (seem-less) [reply forc/src/cli/commands/build.rs:15]: whoops.. missed this

**Review by mitchmindtree — APPROVED** [review]
Great stuff @seem-less, Lgtm :)

**Review by otrho — APPROVED** [review]

**Review by mohammadfawaz — APPROVED** [review]
Excellent!

**Pull request merged**